src/zerialize.ts

```diff
@@ -306,62 +315,94 @@ const zerializers = {
     ...(def.maxLength === null ? {} : { maxLength: def.maxLength.value }),
   }),
 
-  ZodObject: (def) => ({
-    type: "object",
+  ZodObject: (def, ctx) => ({
+    type: 'object',
     properties: Object.fromEntries(
       Object.entries(def.shape()).map(([key, value]) => [
         key,
-        s(value as ZodTypes),
+        s(value as ZodTypes, ctx),
       ])
     ),
   }),
-  ZodRecord: (def) => ({
-    type: "record",
-    key: zerialize(def.keyType),
-    value: s(def.valueType),
+  ZodRecord: (def, ctx) => ({
+    type: 'record',
+    key: s(def.keyType, ctx),
+    value: s(def.valueType, ctx),
   }),
-  ZodMap: (def) => ({
-    type: "map",
-    key: s(def.keyType),
-    value: s(def.valueType),
+  ZodMap: (def, ctx) => ({
+    type: 'map',
+    key: s(def.keyType, ctx),
+    value: s(def.valueType, ctx),
   }),
 
-  ZodEnum: (def) => ({ type: "enum", values: def.values }),
+  ZodEnum: (def) => ({ type: 'enum', values: def.values }),
   // TODO: turn into enum
-  ZodNativeEnum: () => ({ type: "unknown" }),
+  ZodNativeEnum: () => ({ type: 'unknown' }),
 
-  ZodUnion: (def) => ({
-    type: "union",
-    options: def.options.map(s),
+  ZodUnion: (def, ctx) => ({
+    type: 'union',
+    options: def.options.map((option) => s(option, ctx)),
   }),
-  ZodDiscriminatedUnion: (def) => ({
-    type: "discriminatedUnion",
+  ZodDiscriminatedUnion: (def, ctx) => ({
+    type: 'discriminatedUnion',
     discriminator: def.discriminator,
-    options: def.options.map(zerialize),
+    options: def.options.map((option) => s(option, ctx)),
   }),
-  ZodIntersection: (def) => ({
-    type: "intersection",
-    left: s(def.left),
-    right: zerialize(def.right),
+  ZodIntersection: (def, ctx) => ({
+    type: 'intersection',
+    left: s(def.left, ctx),
+    right: s(def.right, ctx),
   }),
 
-  ZodFunction: (def) => ({
-    type: "function",
-    args: zerialize(def.args),
-    returns: zerialize(def.returns),
+  ZodFunction: (def, ctx) => ({
+    type: 'function',
+    args: s(def.args, ctx),
+    returns: s(def.returns, ctx),
   }),
-  ZodPromise: (def) => ({ type: "promise", value: zerialize(def.type) }),
+  ZodPromise: (def, ctx) => ({ type: 'promise', value: s(def.type, ctx) }),
 
-  ZodLazy: (def) => zerialize(def.getter()),
-  ZodEffects: (def) => zerialize(def.schema),
-  ZodBranded: (def) => zerialize(def.type),
-  ZodPipeline: (def) => zerialize(def.out),
-  ZodCatch: (def) => zerialize(def.innerType),
+  ZodLazy: (def, ctx) => {
+    if (ctx.seenSchemas.has(def.getter())) {
+      return { type: 'lazy', ref: ctx.seenSchemas.get(def.getter()) };
```

Comment:
I think in order to remain JSON-compatible in the serialized format, this would need to be a JSON Pointer or such.

Reply:
Updated to use refs.

```json
{
  "type": "object",
  "properties": {
    "limit": { "type": "number", "coerce": true, "id": 3, "isOptional": true },
    "offset": { "type": "number", "coerce": true, "id": 5, "isOptional": true },
    "orderBy": {
      "type": "array",
      "element": {
        "type": "object",
        "properties": {
          "id": {
            "type": "object",
            "properties": {
              "isAsc": {
                "type": "boolean",
                "coerce": true,
                "id": 12,
                "isOptional": true
              },
              "isDesc": {
                "type": "boolean",
                "coerce": true,
                "id": 14,
                "isOptional": true
              }
            },
            "id": 10,
            "isOptional": true
          },
          "test": { "type": "ref", "ref": 10 }
        },
        "id": 9
      },
      "id": 7,
      "isOptional": true
    },
    "id": {
      "type": "object",
      "properties": {
        "isNull": {
          "type": "boolean",
          "coerce": true,
          "id": 18,
          "isOptional": true
        },
        "isNotNull": {
          "type": "boolean",
          "coerce": true,
          "id": 20,
          "isOptional": true
        },
        "eq": {
          "type": "string",
          "coerce": true,
          "id": 22,
          "isOptional": true
        },
        "ne": {
          "type": "string",
          "coerce": true,
          "id": 24,
          "isOptional": true
        },
        "gt": {
          "type": "string",
          "coerce": true,
          "id": 26,
          "isOptional": true
        },
        "gte": {
          "type": "string",
          "coerce": true,
          "id": 28,
          "isOptional": true
        },
        "lt": {
          "type": "string",
          "coerce": true,
          "id": 30,
          "isOptional": true
        },
        "lte": {
          "type": "string",
          "coerce": true,
          "id": 32,
          "isOptional": true
        },
        "like": {
          "type": "string",
          "coerce": true,
          "id": 34,
          "isOptional": true
        },
        "notLike": {
          "type": "string",
          "coerce": true,
          "id": 36,
          "isOptional": true
        },
        "ilike": {
          "type": "string",
          "coerce": true,
          "id": 38,
          "isOptional": true
        },
        "notIlike": {
          "type": "string",
          "coerce": true,
          "id": 40,
          "isOptional": true
        },
        "between": {
          "type": "object",
          "properties": {
            "lower": { "type": "string", "coerce": true, "id": 44 },
            "upper": { "type": "ref", "ref": 44 }
          },
          "id": 42,
          "isOptional": true
        },
        "notBetween": {
          "type": "object",
          "properties": {
            "lower": { "type": "string", "coerce": true, "id": 47 },
            "upper": { "type": "ref", "ref": 47 }
          },
          "id": 45,
          "isOptional": true
        }
      },
      "id": 16,
      "isOptional": true
    },
    "test": { "type": "ref", "ref": 16 },
    "and": {
      "type": "array",
      "element": {
        "type": "lazy",
        "schema": {
          "type": "lazy",
          "schema": {
            "type": "object",
            "properties": {
              "id": { "type": "ref", "ref": 16 },
              "test": { "type": "ref", "ref": 16 },
              "and": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "schema": { "type": "ref", "ref": 51 },
                  "id": 55
                },
                "id": 53,
                "isOptional": true
              },
              "or": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "ref": { "type": "ref", "ref": 51 },
                  "id": 58
                },
                "id": 56,
                "isOptional": true
              }
            },
            "id": 52
          },
          "id": 51
        },
        "id": 50
      },
      "id": 48,
      "isOptional": true
    },
    "or": {
      "type": "array",
      "element": {
        "type": "lazy",
        "ref": {
          "type": "lazy",
          "schema": {
            "type": "object",
            "properties": {
              "id": { "type": "ref", "ref": 16 },
              "test": { "type": "ref", "ref": 16 },
              "and": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "schema": { "type": "ref", "ref": 51 },
                  "id": 55
                },
                "id": 53,
                "isOptional": true
              },
              "or": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "ref": { "type": "ref", "ref": 51 },
                  "id": 58
                },
                "id": 56,
                "isOptional": true
              }
            },
            "id": 52
          },
          "id": 51
        },
        "id": 61
      },
      "id": 59,
      "isOptional": true
    }
  },
  "id": 1
}

```

Reply:
Thanks for the response. However, I think part of the appeal of allowing a JSON serialization format is that it should be easy to craft from JSON as well, so, imo, I really think JSON references would be a much better solution.

Reply:
In case you change your mind let me know. I updated the test and docs.

Reply:
Not my call. Will see what @Gregoor decides...